File: objc/Demo Objc/Controller/SelectPinpadViewController.m
//
//  ConnectPinpadViewController.m
//  Demo Objc
//
//  Created by Eduardo Mello de Macedo | Stone on 02/03/17.
//  Copyright © 2017 Eduardo Mello de Macedo | Stone. All rights reserved.
//

#import "SelectPinpadViewController.h"
#import "NSString+Utils.h"

@implementation SelectPinpadViewController

#pragma mark - Lifecycle

- (void)viewDidLoad {
    // Setup UI components
    [self setupView];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark -Buttons actions

// Action to find connected pinpads
- (IBAction)refresh:(id)sender {
    [self findConnectedPinpads];
}

// Update connected pinpads list and refresh table view content
-(void)findConnectedPinpads {
    _connectedPinpads = [[STNPinPadConnectionProvider new] listConnectedPinpads];
    [self refreshTableViewContent];
}

#pragma mark - UITableViewDataSource

// Set number of rows based on the numbers of available connected pinpads
- (NSInteger) tableView:(nonnull UITableView *)tableView
  numberOfRowsInSection:(NSInteger)section {
    return _connectedPinpads.count;
}

// Set a cell for pinpad from list
- (nonnull UITableViewCell *) tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"pinpadCell" forIndexPath:indexPath];
    STNPinpad *pinpad = _connectedPinpads[indexPath.row];
    cell.textLabel.text = pinpad.name;
    return cell;
}

#pragma mark - UITableViewDelegate

// Connect to the selected pinpad
-(void)tableView:(UITableView *) tableView
    didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    // Check if the row is a valid position at list of connected pinpads
    if ([_connectedPinpads count] > indexPath.row) {
        // The selected pinpad could be retrieved as below
        STNPinpad *pinpad = _connectedPinpads[indexPath.row];
        
        // selectPinpad will try to select the choosed pinpad
        // the return must be used to check connectivity
        [[STNPinPadConnectionProvider new] selectPinpad:pinpad
                                              withBlock:^(BOOL succeeded, NSError * _Nonnull error) {
            NSString *labelContent = [NSString stringWithFormat:@"pinpad %@", pinpad.name];
            if (succeeded) {
                NSLog(@"Pinpad selection succeeded: %@", pinpad);
                labelContent = [@"Valid " stringByAppendingString:labelContent];
            } else {
                NSLog(@"Error: %@", error.description);
                labelContent = [@"Invalid " stringByAppendingString:labelContent];
            }
            // Refresh label data
            [self setFeedbackMessage:labelContent];
        }];
    }
}

#pragma mark - UI Update

// Setup view
-(void)setupView {
    [super viewDidLoad];
    
    self.navigationItem.title = [kTitleSelection localize];
    _instructionLabel.text = [kInstructionSelection localize];
    [_refreshButton setTitle:[kButtonRefresh localize]
                        forState:UIControlStateNormal];
    _overlayView = [[UIView alloc] initWithFrame:[UIScreen mainScreen].bounds];
    _overlayView.backgroundColor = [UIColor colorWithRed:0
                                                       green:0
                                                        blue:0
                                                       alpha:0.5];
    _activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
    _activityIndicator.center = _overlayView.center;
    [_tableView setAllowsMultipleSelection:NO];
    
    // Get the currently selected pinpad.
    STNPinpad *selectedPinpad = [[STNPinPadConnectionProvider new] selectedPinpad];
    
    // Update connectedPinpads
    [self findConnectedPinpads];
    
    // if the selected pinpad exists update the selected row at table view
    if (selectedPinpad != nil && [_connectedPinpads containsObject:selectedPinpad]) {
        int i = (int)[_connectedPinpads indexOfObject:selectedPinpad];
        [_tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:i inSection:0]
                                animated:false
                          scrollPosition:UITableViewScrollPositionTop];
        
        // Update label text with the name of selected pinpad
        _feedback.text = [NSString stringWithFormat:@"Selected pinpad %@", selectedPinpad.name];
    }
}

// Update UI Element
-(void)setFeedbackMessage:(NSString*)message {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.feedback.text = message;
    });
}

// Refresh Table View
-(void) refreshTableViewContent {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.tableView reloadData];
    });
}
@end

File: swift/Demo Swift/SelectPinpadViewController.swift
//
//  ConnectPinpadViewController.swift
//  Demo Swift
//
//  Created by Eduardo Mello de Macedo | Stone on 15/03/17.
//  Copyright © 2017 Eduardo Mello de Macedo | Stone. All rights reserved.
//

import UIKit
import StoneSDK

class SelectPinpadViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    @IBOutlet weak var feedbackLabel: UILabel!
    @IBOutlet weak var tableView: UITableView!
    
    var loadingView: LoadingView!
    
    var connectedPinpads: [STNPinpad]!
    var pinpadConnectionProvider = STNPinPadConnectionProvider()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.navigationItem.title = "Sessão com Pinpad"
        
        //Set loading overlay view
        self.loadingView = LoadingView.init(frame: UIScreen.main.bounds)
        self.navigationController?.view.addSubview(self.loadingView)
        
        tableView.delegate = self
        tableView.dataSource = self
        tableView.allowsMultipleSelection = false
        
        connectedPinpads = []
        self.loadingView.show()
        
        DispatchQueue.main.async()
        {
            self.findConnectedPinpads()
            
            let selectedPinpad = STNPinPadConnectionProvider().selectedPinpad()
            var i = 0
            for pinpad in self.connectedPinpads {
                if selectedPinpad != nil && pinpad.name == selectedPinpad?.name
                {
                    self.tableView.selectRow(at: IndexPath.init(row: i, section: 0),
                                        animated: false,
                                        scrollPosition: UITableViewScrollPosition.top)
                    i += 1
                    self.feedbackLabel.text = "Selected pinpad \(pinpad.name)"
                }
            }
            self.loadingView.hide()
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    func findConnectedPinpads () {
        connectedPinpads = STNPinPadConnectionProvider().listConnectedPinpads()
        for pinpad in connectedPinpads {
            print("Pinpad name: \(pinpad.name) Pinpad identifier: \(pinpad.identifier)")
        }
        print(self.connectedPinpads);
        tableView.reloadData()
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "pinpadCell",
                                                 for: indexPath)
        let pinpad = connectedPinpads[indexPath.row]
        cell.textLabel?.text = pinpad.name
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let selectedPinpad = connectedPinpads[indexPath.row]
        
        
        for i in 0..<1
        {
            self.loadingView.show()
            
            if self.pinpadConnectionProvider.isPinpadConnected(selectedPinpad) == false
            {
                self.pinpadConnectionProvider.connect(to: selectedPinpad)
            }
            self.pinpadConnectionProvider.select(selectedPinpad)
            if self.pinpadConnectionProvider.isPinpadConnected(selectedPinpad)
            {
                DispatchQueue.main.async()
                    {
                        self.feedbackLabel.text = "Selected pinpad \(selectedPinpad.name) - \(selectedPinpad.identifier)"
                }
            }
            else
            {
                DispatchQueue.main.async()
                    {
                        self.feedbackLabel.text = "Unable to select pinpad."
                }
            }
            print("Try to select device number: \(i)")
            self.loadingView.hide()
        }
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return connectedPinpads.count
    }
    
    
    @IBAction func findPinpadsAndRefreshList(_ sender: Any) {
        self.loadingView.show()
        
        DispatchQueue.main.async()
        {
            self.findConnectedPinpads()
            self.loadingView.hide()
        }
    }
}
